doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,522 @@
+```python
+class AbstractPage:
+    def get(self) -> List[EntityReferences]:
+        ...
+    def next(self, pageSize):
+        ...
+
+def relatedEntities(
+    input: EntityReference, relationship) -> AbstractPage: ...
+```
+
+#### Pros
+
+- Simple direct interface for the host application.
+- Simpler for manager plugin implementors to reason about (as evidenced
+  by ad-hoc discussions with collaborators).
+- Hosts should usually have a good idea whether batching or paging is
+  more appropriate for the query they are making.
+- Manager and host do not need to consider both paging and batching at
+  the same time, which is a complex convergence of concepts.
+- Manager implementation can typically be a lightweight wrapper around
+  common HTTP/DB API pagination patterns.
+
+#### Cons
+
+- Precludes batching optimisations by the manager when paging is
+  required.
+- Requires implementation of two methods by the manager: one for
+  batched, one for paged.
+- Potential difficulty for some hosts/queries to determine a priori
+  which API (paging or batch) is the most optimal choice.
+- Having the choice adds a foot-gun for host developers to shoot
+  themselves with by accident.
+- Precludes hosts atomically expressing more complex queries that may
+  be required by spreadsheet-type UIs.
+
+### Batched paging option 2 - Whole-batch paging
+
+E.g. Page token
+
+```python
+# Callback taking index of batch element and selected page of results.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference], relationship,
+    successCb: SuccessCallback, errorCb,
+    limit=-1, pageToken=None
+) -> AbstractPageToken: ...
+```
+
+E.g. Cursor object
+
+```python
+class AbstractPage:
+    def get(self) -> List[List[EntityReference]]:
+        ...
+    def next(self, pageSize):
+        ...
+
+def relatedEntities(
+    batch: List[EntityReference], relationship) -> AbstractPage: ...
+```
+
+#### Pros
+
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Conceptually simpler API for both host and manager if only a single
+  page is active across the whole batch.
+
+### Pros (page token specific)
+
+- Per-element callback means it's trivial to skip processing elements of
+  the input batch whose pages are exhausted.
+- Retains existing design in line with OpenAssetIO design principles,
+  i.e. non-paged signature is simply the same function but ignoring
+  optional arguments.
+
+#### Cons
+
+- Wasted processing of empty lists by the host when pages are exhausted
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+- Batch elements cannot enter an error state in isolation, since paging
+  is whole-batch.
+- Collaborators working on manager plugins raised this as being
+  particularly tricky to implement.
+
+#### Cons (page token specific)
+
+- Likely that the manager plugin must combine multiple back-end page
```

Comment:
[question] This is true of the cursor approach too right? Just they're aggregated in the cursor object not the token object? Or is this really touching on the topic of is a token a value or an object?

Reply:
:thinking: It came from considering REST APIs that use page tokens and assuming that ultimately (whether in the plugin or some back-end) the batch request will actually be several requests with different page tokens, so a single uber page token must be a concatenation of some kind.

But like you say, if that was the back-end behaviour then wrapping in a cursor object doesn't solve the problem.

I'll "promote" that Con to a general problem with whole-batch paging.

Reply:
Moved in cef989e7